src/cargo/core/compiler/job_queue.rs

```diff
@@ -128,24 +233,30 @@ impl<'a> JobState<'a> {
             .tx
             .send(Message::Finish(self.id, Artifact::Metadata, Ok(())));
     }
+
+    /// The rustc underlying this Job is about to acquire a jobserver token (i.e., block)
+    /// on the passed client.
+    ///
+    /// This should arrange for the passed client to eventually get a token via
+    /// `client.release_raw()`.
+    pub fn will_acquire(&self, client: &Client) {
```

Comment:
On passing in the `Client` here, I think it'd still be best to avoid this. I don't think it'll be too hard though with some small tweaks perhaps? Can you try doing something like:

* Add a `Map<Unit, Client>` to `Context`
* Populate this while preparing rustc invocations (only if the jobserver-per-rustc flag is set)
* Internally here we have a map of `Unit` to `JobId`
* This means that given a `JobId` that needs a token, we can get a `Unit`, which we can then use to look up a `Client`